omegatrade-app/backend/app/models/simulation.model.js

```diff
@@ -0,0 +1,83 @@
+'user strict';
+const database = require('./../config/database.js');
+const { v4: uuidv4 } = require('uuid');
+const Simulation = function () { };
+
+Simulation.getAll = async function (cb) {
+    try {
+        const [result] = await database.run({ sql: 'select sml.sId as sId,sml.companyId as companyId, sml.status as status, cy.companyName as companyName, cy.companyShortCode as companyShortCode from simulations sml LEFT JOIN companies cy ON sml.companyId = cy.companyId LIMIT 3', json: true });
+        cb(null, result)
+    } catch (error) {
+        cb(error, null)
+    }
+}
+
+Simulation.findById = async function (param, cb) {
+    try {
+        const sId = param.sId
+        const [ result ] = await database.run({
+            sql: 'select sId,companyId,companyName,companyShortCode,status from simulations where sId = @sId',
+            params: {
+                sId: sId
+            },
+            json: true
+        });
+        cb(null, result)
+        
+    } catch (error) {
+        cb(error, null)
+    }
+}
+
+Simulation.findByCompanyId = async function (companyId, sid) {
+    try {
+        const [result] = await database.run({
+            sql: 'select sId,companyId,companyName,companyShortCode,status from simulations where companyId = @companyId and sid = @sid',
+            params: {
+                companyId: companyId,
+                sid: sid
+            },
+            json:true
+        });
+        return result;
+    } catch (error) {
+        throw ("ERROR:", error);
+    }
+}
+
+Simulation.create = async function (companyId) {
+    try {
+        const sId = uuidv4();
+        await database.table('simulations').insert({
+            sId: sId,
+            status: true,
+            createdAt: 'spanner.commit_timestamp()',
+            companyId: companyId,
+        });
+        return sId;
+    } catch (error) {
+        throw ("ERROR:", error);
+    }
+};
+
+Simulation.deleteById = async function (sId, cb) {
```

Comment:
Here and elsewhere: Why does the `Simulation.deleteById` accept a callback that is called with the result/error of the function, while `Simulation.create` returns / throws its result?

Reply:
@olavloite 
I would like to remove the callback functions in models and the callers in controllers
For example  
in simulation controller

> try{
const simulation = await findById(simulationId);
return res.status(200).json({ success: true, data:simulation });
}catch (error){
     return res.status(500).json({ success: false, message: error });
}

and in simulation model

> simulation.findById = async function(simulationId){
	try	{const sId = param.sId
        const [ result ] = await database.run({
            sql: 'select sId,companyId,companyName,companyShortCode,status from simulations where sId = @sId',
            params: {
                sId: simulationId
            },
            json: true
        });
	return result;
        },catch(error){
        throw Error("Error finding a company by id")  // third suggestion
        }
};

In this approach, All Errors will be handled from a single try-catch block of the controller and any incoming error from the data layer can be given in the response. 
In case of the custom validation message (where I should not return technical information to frontend) will return messages like `user already exists` in case of the already registered user and genric message like `Authentication failed` in case of any failure in the authentication.
And for Debugging the error in order to know the technical information, I've planned to integrate logger service.
please suggest

Reply:
This issue is resolved by removing the callback and try-catch block and used an async function to return the result.
Any error in the data layer will bubble up in the controller catch block and the error log will be logged in the logger service.